File: apis/csharp/routine-api/StoreRoutine.cs
using System;
using System.Collections.Generic;
using System.IO;
using routine_api.exception;
using Thrift;

namespace routine_api
{
    /// <inheritdoc />
    /// <summary>
    /// Defines the base functionality of a Store Routine.
    /// </summary>
    public abstract class StoreRoutine : AbstractRoutine
    {
        private string ProgramId { get; set; }
        private string JobId { get; set; }
        private string TaskId { get; set; }

        /// <summary>
        /// Stores a key/value pair and returns a ResultInfo object. This method contains the StoreRoutine implementation.
        /// </summary>
        /// <param name="partition">The partition id to store to</param>
        /// <param name="key">The key to store</param>
        /// <param name="data">The value to store</param>
        /// <param name="tupleSeq">The current tuple seq #</param>
        /// <returns>Returns a Result object</returns>
        protected abstract Result Store(string partition, string key, byte[] data, int tupleSeq);

        /// <inheritdoc />
        protected override void RunRoutine()
        {
            try
            {
                LinkedList<Result> results = new LinkedList<Result>();

                Log(LogLevel.Debug, "Try to receive from PartitionRoutine");
                // fetch nr of tuples
                int tuples = InPipe.ReadInt();
                Log(LogLevel.Debug, $"{tuples} Tuples to process.");

                for (int i = 0; i < tuples; i++)
                {
                    // receive partition-id, key and value from in stream
                    string partition = InPipe.ReadString();
                    string key = InPipe.ReadString();
                    int size = InPipe.ReadInt();
                    byte[] data = InPipe.ReadBytes(size);

                    // store key value pair
                    Result result = Store(partition, key, data, i);
                    results.AddLast(result);
                    Log(LogLevel.Debug,
                        $"Stored {i + 1}/{tuples}: key {key} and {size} data bytes to partition {partition}");
                }

                // send ResultInfos to RoutineCommunicator (Worker)
                Log(LogLevel.Debug, $"Send {tuples} StoreResult object to RoutinesCommunicator (Worker)");
                Order order = new Order
                {
                    Command = Command.SEND_RESULT,
                    Value = tuples.ToString()
                };
                CtrlPipe.Store(order);
                foreach (Result result in results)
                {
                    CtrlPipe.Store(result);
                }
            }
            catch (Exception e) when (e is TException || e is IOException)
            {
                Log(LogLevel.Error, $"Error while running StoreRoutine: {e.Message}");
                throw new RoutineException(e);
            }
        }

        /// <inheritdoc />
        /// <summary>
        /// Performs a StoreRoutine specific parsing of parameters and setup.
        /// </summary>
        protected override void Setup(string[] args)
        {
            if (args.Length < 7 || args.Length > 8)
            {
                throw new RoutineException("Wrong Number of Arguments for Storage Routine. Usage: <routineName> <inPipe> <ctrlPipe> <pId> <jId> <tId> [<configFile>]");
            }

            string inPipeName = args[2];
            string outPipeName = null;
            string ctrlPipeName = args[3];
            string pId = args[4];
            string jId = args[5];
            string tId = args[6];
            string configFile = null;
            if (args.Length == 8)
            {
                configFile = args[7];
            }

            InPipeName = "null".Equals(inPipeName, StringComparison.InvariantCultureIgnoreCase)
                ? null
                : inPipeName;
            OutPipeName = null;
            CtrlPipeName = "null".Equals(ctrlPipeName, StringComparison.InvariantCultureIgnoreCase)
                ? null
                : ctrlPipeName;

            ProgramId = pId;
            JobId = jId;
            TaskId = tId;

            if (configFile != null)
            {
                Log(LogLevel.Debug, $"Configure StoreRoutine {RoutineName} with {configFile}");
                Configure(configFile);
            }

            SetupStorage();
        }

        /// <inheritdoc />
        /// <summary>
        /// Call the ShutdownStorage method.
        /// </summary>
        protected override void End()
        {
            ShutdownStorage();
        }

        /// <summary>
        /// Configures this routine with a given config file. Method to be implemented by the routine implementation. 
        /// </summary>
        /// <param name="configFile">The path to the config file</param>
        protected abstract void Configure(string configFile);

        /// <summary>
        /// Shutdown any resources opened previously (e.g. storage streams, database connections, ...).
        /// Method to be implemented by the routine implementation. 
        /// </summary>
        protected abstract void ShutdownStorage();

        /// <summary>
        /// Setup any resources needed for the storage routine to work (e.g. storage streams, database connections, ...).
        /// Method to be implemented by the routine implementation. 
        /// </summary>
        protected abstract void SetupStorage();
    }
}

File: routine-api/src/main/java/at/enfilo/def/routine/StoreRoutine.java
package at.enfilo.def.routine;

import at.enfilo.def.routine.api.Command;
import at.enfilo.def.routine.api.Order;
import at.enfilo.def.routine.api.Result;
import org.apache.thrift.TException;

import java.io.IOException;
import java.util.LinkedList;
import java.util.List;

public abstract class StoreRoutine extends AbstractRoutine {

	/**
	 * Store key / value pair and returns a ResultInfo object.
	 *
	 * @param key - key
	 * @param data - value
	 * @param tupleSeq - current tuple seq#
	 * @throws IOException
	 */
	protected abstract Result store(String key, byte[] data, int tupleSeq) throws IOException;


	@Override
	public void runRoutine() {
		try {
			List<Result> results = new LinkedList<>();

			log(LogLevel.DEBUG, "Try to receive from PartitionRoutine");
			// fetch nr of tuples
			int tuples = in.readInt();
			log(LogLevel.DEBUG, String.format("%d Tuples to process.", tuples));

			for (int i = 0; i < tuples; i++) {
				// receive key and value from in stream
				String key = in.readString();
				int size = in.readInt();
				byte[] data = in.readBytes(size);

				// store key value pair
				Result result = store(key, data, i);
				results.add(result);
				log(LogLevel.DEBUG, String.format("Stored %d/%d: key %s and %d data bytes.", i + 1, tuples, key, size));
			}

			// send ResultInfos to RoutineCommunicator (Worker)
			log(LogLevel.DEBUG, String.format("Send %d StoreResult object to RoutinesCommunicator (Worker)", tuples));
			Order order = new Order(Command.SEND_RESULT, Integer.toString(tuples));
			ctrl.store(order);
			for (Result result : results) {
				ctrl.store(result);
			}


		} catch (TException | IOException e) {
			log(LogLevel.ERROR, String.format("Error while running StoreRoutine: %s", e.getMessage()));
			throw new RoutineException(e);
		}
	}


	/**
	 * Configure this routine with the given configuration file.
	 *
	 * @param configFile - configuration file
	 */
	protected abstract void configure(String configFile);


	/**
	 * Shutdown storage IO.
	 *
	 * @throws IOException
	 */
	protected abstract void shutdownStorage() throws IOException;


	/**
	 * Setup storage IO.
	 *
	 * @throws IOException
	 */
	protected abstract void setupStorage() throws IOException;


	/**
	 * StoreRoutine main entry paint.
	 * @param args
	 */
	public static void main(String[] args) {
		if ((args.length != 3) && (args.length != 4)) {
			String errMsg = "3/4 Arguments needed: <routineName> <inPipe> <ctrlPipe> [<configFile>]";
			throw new RoutineException(errMsg);

		} else {

			// Parse arguments
			int arg = 0;
			String routineName = args[arg++];
			String inPipeName = args[arg++];
			String ctrlPipeName = args[arg++];
			String configFile = null;
			if (args.length > arg) {
				configFile = args[arg];
			}

			StoreRoutine routine;
			try {
				// Instantiate routine
				routine = StoreRoutine.class.cast(
						createRoutine(routineName, inPipeName, null, ctrlPipeName)
				);

				// Setup program, job and task infos

				if (configFile != null) {
					routine.log(LogLevel.DEBUG, String.format("Configure StoreRoutine %s with %s", routineName, configFile));
					routine.configure(configFile);
				}

				// Setup specific IO
				routine.setupStorage();

				routine.run();

				// Shutdown specific IO
				routine.shutdownStorage();

			} catch (Exception e) {
				System.err.println(String.format("%s - Error while running StoreRoutine: %s", LogLevel.ERROR, e.getMessage()));
				throw new RoutineException(e);
			}
		}
	}

}

